tests/test_config_flow.py

```diff
@@ -0,0 +1,97 @@
+"""Test the TrueNAS config flow."""
+from unittest.mock import patch
+
+from custom_components.truenas.config_flow import CannotConnect, InvalidAuth
+from custom_components.truenas.const import DOMAIN
+from homeassistant import config_entries, setup
+from websockets.exceptions import InvalidURI, SecurityError
+
+
+async def test_form(hass):
+    """Test we get the form."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == "form"
+    assert result["errors"] == {}
+
+    with patch(
+        "custom_components.truenas.config_flow.Machine.create",
+    ) as mock_machine, patch(
+        "custom_components.truenas.async_setup", return_value=True
+    ) as mock_setup, patch(
+        "custom_components.truenas.async_setup_entry",
+        return_value=True,
+    ) as mock_setup_entry:
+
+        mock_machine.return_value.get_system_info.return_value = {
+            "hostname": "somehostname"
+        }
+
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == "create_entry"
+    assert result2["title"] == "somehostname"
+    assert result2["data"] == {
+        "host": "1.1.1.1",
+        "username": "test-username",
+        "password": "test-password",
+        "name": "TrueNAS",
+    }
+    assert len(mock_setup.mock_calls) == 1
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_invalid_auth(hass):
+    """Test we handle invalid auth."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "custom_components.truenas.config_flow.Machine.create",
+        side_effect=SecurityError,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "invalid_auth"}
+
+
+async def test_form_cannot_connect(hass):
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "custom_components.truenas.config_flow.Machine.create",
+        side_effect=InvalidURI("invalid_uri"),
```

Comment:
The full list of exceptions from the library can be seen [here](https://websockets.readthedocs.io/en/stable/api.html#module-websockets.exceptions).  This is probably good enough, but we might want to handle `InvalidHandshake` as well.

Reply:
In my cursory look, a lot of the errors indicated from the library under `InvalidHandshake` look to be errors that are more fundamental like invalid header and should probably be raised as is.  I say let's keep it as is for now.